core/ledger/kvledger/unjoin_channel.go

```diff
@@ -11,15 +11,58 @@ import (
 	"github.com/hyperledger/fabric-protos-go/ledger/rwset/kvrwset"
 	"github.com/hyperledger/fabric-protos-go/peer"
 	"github.com/hyperledger/fabric/common/ledger/blkstorage"
+	"github.com/hyperledger/fabric/common/ledger/util/leveldbhelper"
 	"github.com/hyperledger/fabric/common/metrics/disabled"
 	"github.com/hyperledger/fabric/core/ledger"
 	"github.com/hyperledger/fabric/core/ledger/confighistory"
 	"github.com/hyperledger/fabric/core/ledger/kvledger/bookkeeping"
 	"github.com/hyperledger/fabric/core/ledger/kvledger/history"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/msgs"
 	"github.com/hyperledger/fabric/core/ledger/kvledger/txmgmt/privacyenabledstate"
 	"github.com/hyperledger/fabric/core/ledger/pvtdatastorage"
+	"github.com/pkg/errors"
 )
 
+// UnjoinChannel removes the data for a ledger and sets the status to UNDER_DELETION.  This function is to be
+// invoked while the peer is shut down.
+func UnjoinChannel(config *ledger.Config, ledgerID string) error {
+	// Ensure the routine is invoked while the peer is down.
+	fileLock := leveldbhelper.NewFileLock(fileLockPath(config.RootFSPath))
+	if err := fileLock.Lock(); err != nil {
+		return errors.Wrap(err, "as another peer node command is executing,"+
+			" wait for that command to complete its execution or terminate it before retrying")
+	}
+	defer fileLock.Unlock()
+
+	// Set the ledger to a pending deletion status.  At the next startup, the peer will complete the removal of any partially deleted ledgers.
+	if err := updateLedgerStatus(config, ledgerID, msgs.Status_UNDER_DELETION); err != nil {
+		return errors.Wrapf(err, "Unjoin channel [%v]", ledgerID)
+	}
+
+	// remove the ledger data
+	if err := removeLedgerData(config, ledgerID); err != nil {
+		return errors.Wrapf(err, "deleting ledger [%v]", ledgerID)
+	}
+
```

Comment:
- After having the data deleted successfully, we need to delete the entry of the ledger from the idStore; leaving the peer in the state that it can join a channel with same name again.
- We would need make change in the boot process so that we continue to delete the partially deleted ledgers, perhaps because of a crash during execution of this commend - as we do for partially constructed [here](https://github.com/hyperledger/fabric/blob/4c4e58c1228968cefc9d5c027f10884fb3707174/core/ledger/kvledger/kv_ledger_provider.go#L122). But this change can be done in a separate PR as it would need it's own tests to make sure the crash consistency of the deletion.